concent_api/core/message_handlers.py

```diff
@@ -1312,27 +1317,42 @@ def handle_send_subtask_results_verify(
             reason=message.concents.ServiceRefused.REASON.InvalidRequest,
         )
 
-    (requestor_has_enough_deposit, provider_has_enough_deposit) = bankster.claim_deposit(
-        concent_use_case=ConcentUseCase.ADDITIONAL_VERIFICATION,
-        requestor_ethereum_address=task_to_compute.requestor_ethereum_address,
-        provider_ethereum_address=task_to_compute.provider_ethereum_address,
-        subtask_cost=task_to_compute.price,
-    )
-
-    if not requestor_has_enough_deposit:
-        return message.concents.ServiceRefused(
-            reason=message.concents.ServiceRefused.REASON.TooSmallRequestorDeposit,
-        )
-    if not provider_has_enough_deposit:
-        return message.concents.ServiceRefused(
-            reason=message.concents.ServiceRefused.REASON.TooSmallProviderDeposit,
-        )
-
     with transaction.atomic(using='control'):
         subtask = get_one_or_none(
             Subtask.objects.select_for_update(),
             subtask_id=compute_task_def['subtask_id'],
         )
+        if subtask is not None and subtask.state_enum in [
+            Subtask.SubtaskState.VERIFICATION_FILE_TRANSFER,
+            Subtask.SubtaskState.ADDITIONAL_VERIFICATION,
+        ]:
+            return message.concents.ServiceRefused(
+                reason=message.concents.ServiceRefused.REASON.DuplicateRequest,
+            )
+
+        (claim_against_requestor, claim_against_provider) = bankster.claim_deposit(
+            subtask_id=task_to_compute.subtask_id,
+            concent_use_case=ConcentUseCase.ADDITIONAL_VERIFICATION,
+            requestor_ethereum_address=task_to_compute.requestor_ethereum_address,
+            provider_ethereum_address=task_to_compute.provider_ethereum_address,
+            subtask_cost=task_to_compute.price,
+            requestor_public_key=requestor_public_key,
+            provider_public_key=provider_public_key,
+        )
+
+        if claim_against_requestor is None:
+            if claim_against_provider is not None:
+                claim_against_provider.delete()
+            return message.concents.ServiceRefused(
+                reason=message.concents.ServiceRefused.REASON.TooSmallRequestorDeposit,
+            )
+        if claim_against_provider is None:
+            if claim_against_requestor is not None:
+                claim_against_requestor.delete()
+            return message.concents.ServiceRefused(
```

Comment:
Why are we removing the claim against requestor's deposit if there's nothing to claim from the provider (and vice versa)? These claims are independent. If bankster says there's nothing to claim from one and does not raise an exception, it should have no influence on the other.

Reply:
:+1: done I thought this is considered as operation failure. After all this stops use case from proceeding. So isn't that claim never finalized?

Reply:
Only `(None, None)` is a failure. And even in this case I now think we should add an additional flag or something to discern failure from the situation where there's simply nothing to pay (I think I suggested this in one of the reviews recently).

If Bankster creates one claim successfully and then fails in a way that does not roll back the transaction automatically, it's supposed to remove the claim and return `(None, None)`.

View should just accept what it gets. If Bankster tells it that only one claim is necessary (and did not fail), then that's it.

Reply:
:+1: done

Reply:
This is not a complete fix. `claim_deposit()` still has a case where it returns only one claim when it wants to indicate a problem (see https://github.com/golemfactory/concent/pull/922/files#r233395518). If you leave it like this then this claim does not get deleted as it should. You must remove it either here or inside `claim_deposit()`. I said that in my opinion the claim should be deleted inside `claim_deposit()`, not here.

Reply:
It's broken now. `claim_deposit()` still has a case where it returns only one claim when it wants to indicate a problem (see https://github.com/golemfactory/concent/pull/922/files#r233395518). If you leave it like this then this claim does not get deleted as it should. You must remove it either here or inside `claim_deposit()`.

I meant that in my opinion the claim should be deleted inside `claim_deposit()`, not that it should not be deleted at all. The function should then return `(None, None)` because it has already cleaned up the mess.

If `claim_deposit()` returns `(None, None)` it might mean either that we don't require payment or that there was not enough deposit. It would be best to add a flag (or maybe an exception?) to make this really explicit. Returning a claim and having the caller delete is a really messy and unintuitive way to indicate this.

Reply:
:+1: done